### PR TITLE
[CALCITE-2900] RelStructuredTypeFlattener generates wrong types on nested columns

### PR DESCRIPTION
**Pull request by my7ym**

### CODE REVIEW

**Comment by my7ym** [comment]
@zabetak Could you please help review the diff when available? Thanks!

**Comment by zabetak** [comment]
I am a little busy these days but I will try to have a look ASAP ;)

File: core/src/main/java/org/apache/calcite/sql2rel/RelStructuredTypeFlattener.java [hunk 1]
```diff
@@ -303,9 +303,11 @@ protected int getNewForOldInput(int oldOrdinal) {
    * corresponding field post-flattening, and also returns its type.
    *
    * @param oldOrdinal Pre-flattening ordinal
+   * @param existingOffset offset already calculated the target column inside the oldOrdinal column.
+   *                       For unnested column, it should be 0.
    * @return Post-flattening ordinal and type
    */
-  protected Ord<RelDataType> getNewFieldForOldInput(int oldOrdinal) {
+  protected Ord<RelDataType> getNewFieldForOldInput(int oldOrdinal, int existingOffset) {
```
Comment (zabetak) [inline core/src/main/java/org/apache/calcite/sql2rel/RelStructuredTypeFlattener.java]: Why protected and not private? Is it meant to be subclassed?

File: core/src/test/java/org/apache/calcite/tools/PlannerTest.java [hunk 2]
```diff
@@ -996,6 +997,66 @@ private void checkBushy(String sql, String expected) throws Exception {
     assertThat(toString(transform), containsString(expected));
   }
 
+  private void checkFlattenTypesOnNestedColumn(String sql, String expected) throws Exception {
```
[the remark this answers is not in the document]
Comment (my7ym) [reply core/src/test/java/org/apache/calcite/tools/PlannerTest.java]: I could change the name checkQueryPlan or some other better names you could think of.

I actually wanted to merge the checkBushy and the new helper together. But actually it may be cleaner to keep them both because 
1. FrameworkConfig is different because default will only identify table `BOOKSTORE` but not `bookstore` and similar things.
2. defaultSchema is different.
3. More importantly, flow is different. checkBushy needs `project`, convert traitSet and transform while my test do not need them.

If I force to merge them, it won't necessarily make it better. That's why I did not merge them.

File: core/src/test/java/org/apache/calcite/tools/PlannerTest.java [hunk 3]
```diff
@@ -996,6 +997,66 @@ private void checkBushy(String sql, String expected) throws Exception {
...
+    assertThat(toString(convert), containsString(expected));
+  }
+
+  /** Test case for
+   * <a href="https://issues.apache.org/jira/browse/CALCITE-2900">[CALCITE-2900]
+   * RewriteRexShuttle generates wrong type during flatten for structured type when
+   * visitInputRef</a>. */
+  @Test
+  public void testFlattenTypesWithFieldAccessOnFirstLevel() throws Exception {
+    String sql = ""
```
[the remark this answers is not in the document]
Comment (my7ym) [reply core/src/test/java/org/apache/calcite/tools/PlannerTest.java]: Thanks for detailed review and good catch! Will make the tests simpler and will try to add low-level tests

File: core/src/test/java/org/apache/calcite/test/BookstoreSchema.java [hunk 4]
```diff
@@ -90,16 +96,50 @@ public Author(int aid, String name, Place birthPlace, List<Book> books) {
   /**
    */
   public static class Place {
-    public final Coordinate coords;
     public final String city;
     public final String country;
+    public final double elevationInMeters;
+    public final Coordinate coords;
+    public final Area area;
```
[the remark this answers is not in the document]
Comment (my7ym) [reply core/src/test/java/org/apache/calcite/test/BookstoreSchema.java]: I tested against elevationInMeters and Area.number.

The reason why they are necessary because the targeted type has to be different than the first field's type of the current level. Here, elevationInMeters type has to be different than city, and area.number has to be different than areaUnit.

File: core/src/test/java/org/apache/calcite/test/BookstoreSchema.java [hunk 5]
```diff
@@ -90,16 +96,50 @@ public Author(int aid, String name, Place birthPlace, List<Book> books) {
   /**
    */
   public static class Place {
-    public final Coordinate coords;
     public final String city;
     public final String country;
+    public final double elevationInMeters;
```
[the remark this answers is not in the document]
Comment (my7ym) [reply core/src/test/java/org/apache/calcite/test/BookstoreSchema.java]: Exactly man.

File: core/src/main/java/org/apache/calcite/sql2rel/RelStructuredTypeFlattener.java (hunk 1 above)
Comment (my7ym) [reply core/src/main/java/org/apache/calcite/sql2rel/RelStructuredTypeFlattener.java]: Hmmm, to your question YES and I did this because in the past it is protected and it's not a final class. So I could not eliminate the possibility that somebody subclasses it.

And thanks man you do remind me that I need to keep the original function signature because someone may subclass it. if there is a @override annotation then this change is not backward compatible.

Comment (my7ym) [reply core/src/main/java/org/apache/calcite/sql2rel/RelStructuredTypeFlattener.java]: Oh, I re-read the code and actually I did not delete the original method. I will make it private since I don't see any points that people will override it. Good catch!

File: core/src/main/java/org/apache/calcite/sql2rel/RelStructuredTypeFlattener.java [hunk 6]
```diff
@@ -303,9 +303,11 @@ protected int getNewForOldInput(int oldOrdinal) {
    * corresponding field post-flattening, and also returns its type.
    *
    * @param oldOrdinal Pre-flattening ordinal
+   * @param existingOffset offset already calculated the target column inside the oldOrdinal column.
+   *                       For unnested column, it should be 0.
    * @return Post-flattening ordinal and type
    */
-  protected Ord<RelDataType> getNewFieldForOldInput(int oldOrdinal) {
+  private Ord<RelDataType> getNewFieldForOldInput(int oldOrdinal, int existingOffset) {
```
Comment (my7ym) [inline core/src/main/java/org/apache/calcite/sql2rel/RelStructuredTypeFlattener.java]: @zabetak Changed to private.

File: core/src/main/java/org/apache/calcite/sql2rel/RelStructuredTypeFlattener.java [hunk 7]
```diff
@@ -328,13 +330,24 @@ protected int getNewForOldInput(int oldOrdinal) {
...
   }
 
+  /**
+   * Maps the ordinal of a field pre-flattening to the ordinal of the
+   * corresponding field post-flattening, and also returns its type.
+   *
+   * @param oldOrdinal Pre-flattening ordinal
+   * @return Post-flattening ordinal and type
+   */
+  protected Ord<RelDataType> getNewFieldForOldInput(int oldOrdinal) {
```
Comment (my7ym) [inline core/src/main/java/org/apache/calcite/sql2rel/RelStructuredTypeFlattener.java]: @zabetak This is the old function signature so I have to keep them to keep backward compatibility.

File: core/src/test/java/org/apache/calcite/test/SqlToRelConverterTest.java [hunk 8]
```diff
@@ -2559,6 +2559,18 @@ public void testDynamicStarInTableJoin() throws Exception {
     sql(sql).with(getTesterWithDynamicTable()).ok();
   }
 
+  /**
+   * Test case for <a href="https://issues.apache.org/jira/browse/CALCITE-2900">[CALCITE-2900]
+   * RewriteRexShuttle generates wrong type during flatten for structured type when
+   * visitInputRef</a>.
+   */
+  @Test
+  public void testNestedColumn() {
```
Comment (my7ym) [inline core/src/test/java/org/apache/calcite/test/SqlToRelConverterTest.java]: @zabetak This is the new a little "low-level" test I added. SqlToRelConverter.flattenTypes is almost the lowest level entry point for this test. The other option is RelStructuredTypeFlattener.rewrite. But for this test, I have to write a lot of boilerplate code for this single test. Since there is almost no extra logic applied in SqlToRelConverter.flattenTypes, I think it should be OK.

Let me know. Thanks!

File: core/src/test/java/org/apache/calcite/tools/PlannerTest.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (zabetak) [reply core/src/test/java/org/apache/calcite/tools/PlannerTest.java]: On a second though why not moving all the tests in SqlToRelConverterTest? Like that you don't need to introduce a separate checkQueryPlan method.

File: core/src/test/java/org/apache/calcite/test/SqlToRelConverterTest.java (hunk 8 above)
Comment (zabetak) [reply core/src/test/java/org/apache/calcite/test/SqlToRelConverterTest.java]: SqlToRelConverterTest is a good place. Actually it seems that you can move all the tests here. Moreover this new test does not bring something significant in comparison with the previous ones so I would suggest to remove it.

File: core/src/test/java/org/apache/calcite/test/catalog/Fixture.java [hunk 9]
```diff
@@ -109,16 +109,23 @@
             .build();
     empListType =
         typeFactory.createArrayType(empRecordType, -1);
+
+    // Subclass ObjectSqlType to make nested field queryable.
     addressType =
-        new ObjectSqlType(SqlTypeName.STRUCTURED,
+        new ObjectSqlType(
+            SqlTypeName.STRUCTURED,
```
Comment (zabetak) [inline core/src/test/java/org/apache/calcite/test/catalog/Fixture.java]: Changing only the BookStore schema should be sufficient to demonstrate what you want. In RelToSqlConverterTest you should only pass the appropriate SchemaSpec before writting your sql.

Comment (my7ym) [reply core/src/test/java/org/apache/calcite/test/catalog/Fixture.java]: Hey @zabetak, I tried this Bookstore schema in SqlToRelConverterTest. The change is much bigger than the original one because to add Bookstore schema in SqlToRelTestBase takes:
1. Change TesterImpl type factory to JavaTypeFactoryImpl -> add withTypeFactory method in both Tester interface and impl.
2. Do the same thing for RelOptPlanner in TesterImpl and replace it with a subclass of MockRelOptPlanner with traitset related method overrided. 
3. To support the Bookstore schema, we need make the CalciteSchema to be cached instead of existing simple, and make it case insensitive -> change MockCatalogReader constructor signature and subclass it to register bookstore schema.

I think there is no need that we stick to Bookstore because there is already existing nested fields in MockCatalogReaderSimple. Another approach is to adapt the existing changes on Bookstore schema to MockCatalogReaderSimple (basically add one more layer deeper in a nested schema). And migrate PlannerTest cases to SqlToRelConverterTest. IMO this approach is cleanest and also aligned with the existing tests.

What do you think?

Comment (my7ym) [reply core/src/test/java/org/apache/calcite/test/catalog/Fixture.java]: @zabetak And if you think OK, we could just keep the existing second-level-access test I implemented in SqlToRelConverterTest. I don't think there is much difference between the second level and the third level.

If so, I will delete all changes made to Bookstore and PlannerTest.

let me know.

Comment (zabetak) [reply core/src/test/java/org/apache/calcite/test/catalog/Fixture.java]: Thanks for your detailed explanation @my7ym! Indeed, I was not expecting it to be so complicated. I agree with you, probably the best place is SqlToRelConverterTest and there is no need to touch the Bookstore schema.

Comment (my7ym) [reply core/src/test/java/org/apache/calcite/test/catalog/Fixture.java]: No problem. Learnt a lot actually and more familiar with the code base :). Pls take another look man. Thanks.

File: core/src/test/java/org/apache/calcite/test/catalog/Fixture.java [hunk 10]
```diff
@@ -109,16 +109,23 @@
...
             false,
             Arrays.asList(
                 new RelDataTypeFieldImpl("STREET", 0, varchar20Type),
                 new RelDataTypeFieldImpl("CITY", 1, varchar20Type),
                 new RelDataTypeFieldImpl("ZIP", 2, intType),
                 new RelDataTypeFieldImpl("STATE", 3, varchar20Type)),
-            RelDataTypeComparability.NONE);
+            RelDataTypeComparability.NONE) {
+          @Override public StructKind getStructKind() {
+            return StructKind.PEEK_FIELDS_DEFAULT;
```
Comment (zabetak) [inline core/src/test/java/org/apache/calcite/test/catalog/Fixture.java]: I guess before it was StructKind.FULLY_QUALIFIED, wasn't this enough to demonstrate the problem?

File: core/src/test/java/org/apache/calcite/test/SqlToRelConverterTest.java [hunk 11]
```diff
@@ -2559,6 +2559,18 @@ public void testDynamicStarInTableJoin() throws Exception {
...
+    final String sql =
+        "select HOME_ADDRESS.ZIP from SALES.EMP_ADDRESS where HOME_ADDRESS.CITY = 'abc'";
```
Comment (zabetak) [inline core/src/test/java/org/apache/calcite/test/SqlToRelConverterTest.java]: Minor: I guess it is just a matter of style but since the majority of tests use lowercase formatting for table and column identifiers maybe it is worth doing the same. Of-course it is not consistent all over the place so I have no strong preference on this.

**Comment by zabetak** [comment]
I changed a bit the title to describe the problem at a higher level. Let me know if what I wrote is not correct. If it looks good change the commit message respectively and and suffix with your name in parentheses.

**Comment by my7ym** [comment]
@zabetak The title is much better. Done with commit message change. Thanks very much for all the detailed and constructive reviews!

File: core/src/test/java/org/apache/calcite/test/catalog/Fixture.java (hunk 10 above)
Comment (my7ym) [reply core/src/test/java/org/apache/calcite/test/catalog/Fixture.java]: Sorry that it was not enough to do that. It will make the nested column unrecognizable by part of the flow. In my personal project I used PEEK_FIELDS_DEFAULT, so I stick to it here.

File: core/src/test/java/org/apache/calcite/test/SqlToRelConverterTest.java (hunk 11 above)
Comment (my7ym) [reply core/src/test/java/org/apache/calcite/test/SqlToRelConverterTest.java]: Good call. Done.

**Comment by my7ym** [comment]
@zabetak Could you help revert this fix? I think there is some corner cases of fixes which will lead to crash. Thanks.
